### ISSUE TITLE
complex LN numerical issue

### ISSUE THREAD

**Issue by achan001** [thomasokken/free42#62]
**complex LN** has numerical issue (both binary/decimal version)

Test : x = (pi + pi*I) / 1e10, evaluate ln(y = 1+x)
Note: log1p(x) ≈ ln(y) - (y-1-x)/y

Binary version reference:
```
p2> from gmpy2 import *
p2> x = const_pi() / 1e10
p2> x = mpc(x,x)
p2> y = 1+x
p2> log(y)
mpc('3.1415936518897064e-10+3.141592652602832e-10j')
```

lua> x = I(pi,pi) / 1e10
lua> I.log(1+x) -- error = -1 ULP
(3.14159365188970**7**e-10+3.141592652602832e-10*I)

Plus42Binary> LN(1+x)  -- error = +954529 ULP
3.141593651**3962265**e-10+3.141592652602832e-10i

Same issue with Plus42Decimal

Y = 1+X = 
1.000000000314159265358979323846264 +
3.141592653589793238462643383279503e-10i

LN(Y), Wolfram Alpha = -- reference
3.141592653589793238255931488895411E-10 +
3.141592652602832798560416032598054E-10i

LN(Y), Plus42Decimal = -- error = -3,452,423,245 ULP
3.14159265358979323825593**4941318656**e-10 +
3.141592652602832798560416032598054e-10i

**Comment by achan001** [comment]
I discovered **complex LN** numerical issue when testing latest function **C.LN1+X**

It is funny I had exactly the same experience when adding complex.log1p
I've had to re-implement complex.log to fix numerical issue.

see https://www.hpmuseum.org/forum/thread-19784-post-171238.html#pid171238

complex.log lua code was later trranslated in C.
```
static Complex Clog(Complex z) {
    double x = Real(z), y = Imag(z);
    double h = x*x + y*y;
    if (0.5 < h && h < 3.0) {
        if (fabs(x) < fabs(y)) {h=x; x=y; y=h;}
        h = log1p((x+1)*(x-1) + y*y)/2;
    } else
        h = log(cabs(z));
    return CMPLX(h, carg(z));
}
```

**Comment by achan001** [comment]
The good news, **C.E^X-1** function looks good.

X =
3.141592653589793238462643383279503e-10 + 
3.141592653589793238462643383279503e-10i

L = LN(Y) - (Y-1-X)/Y, using WA accurate LN(Y)
3.141592653589793238255934872174913e-10 + 
3.141592652602832798560416032598053e-10i

C.E^X-1(L) = -- round-trip back to X :-)
3.141592653589793238462643383279503e-10 + 
3.141592653589793238462643383279503e-10i

**Comment by thomasokken** [comment]
I implemented your fix for complex LN, and applied the same code for complex LOG as well (using LOG = LN / LN(10)): dd501ebc4e0cf5dd3b759d82b47a686f5faff8c3

Doing the same for C.LN1+X is going to be a bit more complicated; I haven't looked into that yet.

**Comment by achan001** [comment]
> Doing the same for C.LN1+X is going to be a bit more complicated

Good **LN** is all it take. 
No changes needed for **C.LN1+X**

y=1+x --> log1p(x) ≈ ln(y) - (y-1-x)/y

**Comment by thomasokken** [comment]
Ah, OK. So that means I shouldn't need any more code changes for this issue, then?
The revised code returns results that match your test cases for (pi+1, pi), binary and decimal.
Let me know if you would like to take a look at a test build (and if so, for which OS). I'll try to include the ATANH fix as well; I'm starting on that right now.

**Comment by achan001** [comment]
It appeared LN has dead code
```
    phloat h = xre * xre + xim * xim;
    phloat s = 0;
    if (p_isinf(h)) {
        s = 10000;
        xre /= s;
        xim /= s;
        h = xre * xre + xim * xim;   // not needed
        s = log(s);
    }
```

If h is infinite, it will never enter (0.5 < h && h < 3.0) branch.
No need for h update (which also will not enter this branch)

Also, s = 100000 may be better scaling factor
For 34-digits decimal, ln(1e5) error = +0.10 ULP vs ln(1e4) = -0.17 ULP
For 53-bits binary, ln(1e5) error = -0.11 ULP vs ln(1e4) error = -0.49 ULP

My preference is to do special case first. But it does not really matter.
```
    phloat h = xre * xre + xim * xim;
    if (0.5 < h && h < 3.0) {
        if (fabs(xre) < fabs(xim))
            h = log1p((xim + 1) * (xim - 1) + xre * xre) / 2;
        else
            h = log1p((xre + 1) * (xre - 1) + xim * xim) / 2;
    } else {
        phloat s = 0;
        if (p_isinf(h)) {
            s = 100000;
            xre /= s;
            xim /= s;
            s = log(s);
        }
        h = log(hypot(xre, xim)) + s;
    }
    *yre = h;
    *yim = atan2(xim, xre);
    return ERR_NONE;
```

**Comment by achan001** [comment]
```
if (fabs(x) < fabs(y)) {h=x; x=y; y=h;}
h = log1p((x+1)*(x-1) + y*y)/2;
```
I've found a more accurate way, (x^2-1) = 2*(x-1) + (x-1)^2

[Clog version 1](https://www.hpmuseum.org/forum/thread-21258-post-183702.html#pid183702)
[Clog version 2](https://www.hpmuseum.org/forum/thread-21258-post-183730.html#pid183730)

**Comment by thomasokken** [comment]
I'm getting back to this issue after having spent a few weeks on other Free42 & Plus42 work.
I see you've posted several times at MoHPC.
Which version would you recommend I use for LN and LOG? Is "Clog version 2" from your last post here still the best?

**Comment by achan001** [comment]
ln(|z|(1+ε)) ≈ ln(|z|) + ε
relerr(|z|) = ε      ⇒ relerr(ln(|z|) ≈ ε / ln(|z|)

If |z| is not close to 1, ln(|z|) can't improve much using **log1p**
In that sense, both version 1 and 2 are equally good.

I use version 2 for its elegance without branching for edge cases.

**Comment by thomasokken** [comment]
The version 2 code appears to use logb() and ldexp() as each other's inverses, but logb() uses base FLT_RADIX and ldexp() uses base 2. Is that correct? It looks like it should use scalbn() instead of ldexp() if it's going to work with decimal floating point, or else it should use log2() instead of logb() if the scaling calculation is supposed to use base 2 even with decimal floating point.

Also, is the shift_ln2() function compatible with decimal floating point?

**Comment by achan001** [comment]
You are correct. **logb** should match with **scalbn**.

complex.log version 2 code assumed FLT_RADIX = 2.
Actually, math library I used also assumed base 2.

https://github.com/JuliaMath/openlibm/blob/master/src/s_logb.c
https://github.com/JuliaMath/openlibm/blob/master/src/s_scalbn.c

To make version 2 work with decimal is more involved.
We also need accurate shift_ln10(x,b) = x + b*ln(10)
Or, do everything in base 2, then convert back to decimal.
It may be too much hassle for not much gain in accuracy.

I would just replace (x+1)*(x-1) ≡ (2ε + ε²), where ε = |x| - 1

**Comment by thomasokken** [comment]
Wouldn't doing everything in base 2 and then converting back to decimal mean getting results with only 16 out of 34 digits being accurate?

**Comment by achan001** [comment]
No.

if |z| not close to 1, dec→bin conversion error does not matter.
if |z| close to 1, ε = dec→bin ( |x|-1 ) should have almost all sig. digits.

Intel Decimal128 internally does dec → bin → dec conversion anyway.
Example, **bid128_log1p.c**, log1p after handled edge cases:
```
   { BIDECIMAL_CALL1(bid128_to_binary128,xd,x);
     __bid_f128_log1p(yd, xd);
     BIDECIMAL_CALL1(binary128_to_bid128,res,yd);
     BID_RETURN(res);
   }
```

**Comment by thomasokken** [comment]
> Intel Decimal128 internally does dec → bin → dec conversion anyway.

Yes, but Free42 Decimal uses 128-bit decimal numbers, so if you do calculations using 64-bit binary double, you end up taking a 16-digit result and presenting it as if it had 34 actual digits of precision.

**Comment by achan001** [comment]
I now understand why you say binary only give 16 out of 34 digits.

When I say do everything in base 2, then convert back to decimal, 
I meant **decimal128** to **binary128** (113 bits ≈ 34 decimal digits)

This is messy (2 versions of complex log), without much gain in accuracy.
That's why I suggested *not* use version 2.

> I would just replace (x+1)*(x-1) ≡ (2ε + ε²), where ε = |x| - 1

**Comment by thomasokken** [comment]
I implemented version 1b: https://github.com/thomasokken/free42/commit/1f9041beaa534b982d68266ac7c945988f9a65b9

There are other changes in that commit as well; the relevant code for LN is near the end, in mappable_ln_c().

**Comment by achan001** [comment]
Looks good.

Same reason **mappable_ln_c()** used 0x1p61 for binary subnormal case,
I would use 1E38 to scale sub-normal decimal case, not 1E37

ln(1E37)  ≈ 85.19564844077969030866568382332148 + -0.432 ULP
ln(1E38)  ≈ 87.49823353377373599268367527800584 + **-0.011 ULP**

**Comment by thomasokken** [comment]
Great, thanks!

I also did the changes for LN and C.LN1+X. I combined everything in one commit: https://github.com/thomasokken/free42/commit/96bcd842188cd410b95c58adbb0847e11684ec62

I uploaded test builds incorporating everything to https://thomasokken.com/free42/download/test/

Note that these do not include your suggestion about using 1E38 for the decimal subnormal case; I'll add that before building the official release.

**Comment by achan001** [comment]
Your code had atanh(z) and atan(z) mixed up.
With swap((x,y)) = (y,x), identity is **atanh(z) = swap(atan(swap(z)))**

With commit: https://github.com/thomasokken/free42/commit/96bcd842188cd410b95c58adbb0847e11684ec62, Free42-Decimal get

```
Z = (PI,PI) / 1E10
ATANH(Z/(2+Z))*2 = 3.141592653589793238359289127678504e-10+3.141592652602832798457061776997054e-10i
ATAN(Z/(2+Z))*2  = 3.141592653589793238255934872174912e-10+3.141592652602832798560416032598052e-10i
C.LN1+X(Z)       = 3.141592653589793238255934872174914e-10+3.141592652602832798560416032598053e-10i
```

We should have **log1p(z) = 2*atanh(z/(2+z))**, not with atan
```
p2> from mpmath import *
p2> mp.dps = 34
p2> mp.pretty = 1
p2> z = mpc(pi,pi) / 1e10
p2> y = z / (2+z)
p2> atan(y)*2
(0.0000000003141592653589793238359289127678504 + 0.0000000003141592652602832798457061776998517j)
p2> atanh(y)*2
(0.0000000003141592653589793238255934872174753 + 0.0000000003141592652602832798560416032598053j)
p2> log1p(z)
(0.0000000003141592653589793238255934872174913 + 0.0000000003141592652602832798560416032598053j)

**Comment by thomasokken** [comment]
Fixed.

I uploaded new test builds, they're in the same location as before, https://thomasokken.com/free42/download/test/

**Comment by thomasokken** [comment]
I released Free42 3.1.6 and Plus42 1.1.8 today, containing these changes.

**Comment by achan001** [comment]
> I would use 1E38 to scale sub-normal decimal case, not 1E37

Turns out, the constant ULP error really doesn't matter.
ln(1E37) ≈ 85, but it is only used when |z|^2 underflow/overflow.

| ln(|z|) | ≥ ln(1E6145) / 2 ≈ 7075

Decimal offset is at least 2 magnitude smaller.

---

BTW, in decimal128, ln(1E38) ≠ ln(10) * 38

10 LN 38 ×      → 87.4982335337737359926836752780058**3**
1E38 LN           → 87.49823353377373599268367527800584

Numerically, ln(a^b) may not equal ln(a) * b
It may be better not to waste the accurate constant.
```
#ifdef BCD_MATH
phloat m = scalbn(phloat(1), 38);
#else
phloat m = 0x2000000000000000; // 2^61
#endif
phloat b = -log(m);
```
